### DISCUSSION TITLE
Docs Improvement on SSL Certificates (GoDaddy Example)

### DISCUSSION THREAD

**Discussion by Lunar2kPS** [verdaccio/verdaccio#3109] [Q&A] [answered]
## My Basic Question:

I went through a lot of trouble setting up an SSL certificate from GoDaddy to work with Verdaccio. I finally found an answer after a week, and was wondering if I could help improve the Verdaccio docs by providing something like my example?

_(Note I'm using docker-compose, so I could provide a helpful example of that too!)_

## Context:
I bought an SSL Certificate from GoDaddy so I could more-easily serve my Verdaccio package registry over HTTPS and have it "just work" on new machines. For example, I use the Unity Package Manager to download my packages, and Postman to test HTTPS requests to the registry to see if it's working.

I kept getting stuck by "unable to verify first certificate" SSL errors in both Unity and Postman.
I didn't want to set my `npm config set strict-ssl false` -- since that disables a lot of security!
I didn't want to manually set my NPM to use a `cafile`, nor setup my Postman to manually add one of the GoDaddy certificates to.. that'd be a lot of setup if everyone's SSL certificate required manual steps!

So I finally found the answer.

## The Main Problem -- 4 files, 3 `config.yaml` fields

The weird thing for me was that I had the following files:
- `private-key.pem`
- `csr.pem`
- `57424864cf1921b1.crt` _(from GoDaddy, in my example case)_
- `gd_bundle-g2-g1.crt` _(from GoDaddy, in my example case)_

The thing that confused me was that there are **3 fields** in the Verdaccio `config.yaml`:
![image](https://user-images.githubusercontent.com/25470751/162329809-d08cbbff-cc5f-454d-a063-9158c361d050.png)

But with 2 certificate files from GoDaddy, I was unsure which to use.

### Small Question About File Extensions:
Please correct me if I'm using the wrong file extensions (by convention) for the keys/certificates/CSRs.
I know they're all text files, but I've seen `.pem`, `.crt`, and sometimes `.key` or `.txt`... so I'm not sure really what should be used where haha..

## Configuring the `config.yaml` file with SSL Certificates...

Here's what I found:
1. The `gd_bundle-g2-g1.crt` file is their "Intermediate Certificate"
2. The `57424864cf1921b1.crt` file is my certificate for my domain, signed by GoDaddy (I got a DV (Domain Validated) SSL Certificate)
3. For the `https.cert` field in `config.yaml`...
    1. If I set it to my `gd_bundle-g2-g1.crt` file, it'd fail to start Verdaccio.
    2. If I set it to my `57424864cf1921b1.crt` file, it'd fail to verify the certificate.
    3. :fire: :tada: HOWEVER, if I created a new empty text file, and copied (order is very important) FIRST **my certificate's text** (`57424864cf1921b1.crt`), one newline character, THEN SECOND **GoDaddy's intermediate certificate's text,** and set `https.cert` to this new combined certificate file... **it WORKS!** All my errors went away, and I can use HTTPS!

---

Also thanks to [this explanation](https://www.digitalocean.com/community/tutorials/how-to-install-an-ssl-certificate-from-a-commercial-certificate-authority#download-certificate) that helped understand the 2 cert files from GoDaddy!

**Comment by juanpicado** [comment, marked as answer]
>  was wondering if I could help improve the Verdaccio docs by providing something like my example?

Absolutely YES 👏🏼

**Reply by Lunar2kPS** [reply]
Awesome!!

I think I have a rough idea on how to start under the `website/docs` folder, so I'll start a PR in the coming days and probably get your (and/or others') feedback to make sure it all looks right :)

**Reply by juanpicado** [reply]
yes, thats the folder, I would go with this page https://verdaccio.org/docs/ssl I think fits better. The file is `ssl.md`

**Reply by Lunar2kPS** [reply]
Oh! I was thinking of making it a child page underneath that page, is that possible?
(The example has a lot of information, so I wouldn't wanna clutter the SSL basics on that page)

**Reply by juanpicado** [reply]
If you want to have enough space, you can have it as this one https://verdaccio.org/docs/aws , new file `gogaddy.md` under the guides , sounds that reasonable?

**Reply by Lunar2kPS** [reply]
Gotcha, yeah that's perfect, thanks!

**Reply by juanpicado** [reply]
Then you can add it to the list here ´website\sidebars.js´ so is displayed on the sidebar correctly

**Reply by Lunar2kPS** [reply]
Oh awesome, I was gonna ask but didn't want to ask too many questions cause I'm sure you're busy haha.
Will-do, excited to make my first contribution! :)

**Reply by Lunar2kPS** [reply]
By the way!
I haven't forgotten about this, I just got really busy this month!

I'm not closing this tab out on my browser until I finally get around to writing up the docs and PR 😆

**Comment by juanpicado** [comment]
> Please correct me if I'm using the wrong file extensions (by convention) for the keys/certificates/CSRs.
I know they're all text files, but I've seen .pem, .crt, and sometimes .key or .txt... so I'm not sure really what should be used where haha..

At this point I can only refer you to the Node.js docs, https://nodejs.org/api/https.html#httpscreateserveroptions-requestlistener because verdaccio uses `https.createServer(httpsOptions, app);`
